metagraph/src/graph/representation/hash/dbg_sshash.cpp

```diff
@@ -0,0 +1,494 @@
+    void visit(essentials::saver &saver) const { dict_.visit(saver); }
+
+    void access(uint64_t kmer_id, char* string_kmer) const { dict_.access(kmer_id, string_kmer); }
+    void build(std::string const& input_filename, sshash::build_configuration const& build_config) {
+        dict_.build(input_filename, build_config);
+    }
+
+    uint64_t lookup(char const* string_kmer, bool check_reverse_complement = true) const {
+        return dict_.lookup(string_kmer, check_reverse_complement);
+    }
+
+    uint64_t lookup_uint(kmer_t uint_kmer, bool check_reverse_complement = true) const {
+        return dict_.lookup_uint(uint_kmer, check_reverse_complement);
+    }
+
+    sshash::lookup_result lookup_advanced(
+            char const* string_kmer,
+            bool check_reverse_complement = true) const {
+        return dict_.lookup_advanced(string_kmer, check_reverse_complement);
+    }
+
+    sshash::lookup_result lookup_advanced_uint(
+            kmer_t uint_kmer,
+            bool check_reverse_complement = true) const {
+        return dict_.lookup_advanced_uint(uint_kmer, check_reverse_complement);
+    }
+
+    sshash::neighbourhood<kmer_t> kmer_forward_neighbours(
+            char const* string_kmer,
+            bool check_reverse_complement = true) const {
+        return dict_.kmer_forward_neighbours(string_kmer, check_reverse_complement);
+    }
+
+    sshash::neighbourhood<kmer_t> kmer_backward_neighbours(
+            char const* string_kmer,
+            bool check_reverse_complement = true) const {
+        return dict_.kmer_backward_neighbours(string_kmer, check_reverse_complement);
+    }
+
+  private:
+    // TODO: this is a hack since sshash::dictionary<kmer_t>::visit is always non-const
+    mutable sshash::dictionary<kmer_t> dict_;
+};
+
+constexpr DeBruijnGraph::node_index sshash_to_graph_index(uint64_t idx) {
+    return idx + 1;
+}
+constexpr uint64_t graph_index_to_sshash(DeBruijnGraph::node_index idx) {
+    return idx - 1;
+}
+
+DBGSSHash::DBGSSHash(size_t k, Mode mode) : k_(k), num_nodes_(0), mode_(mode) {
+    size_t odd_k = k_ % 2 == 0 ? k_ + 1 : k_;
+
+    if (odd_k * IDictionary::bits_per_char <= 64) {
+        dict_ = std::make_unique<Dictionary<Kmer64>>();
+    } else if (odd_k * IDictionary::bits_per_char <= 128) {
+        dict_ = std::make_unique<Dictionary<Kmer128>>();
+    } else if (odd_k * IDictionary::bits_per_char <= 256) {
+        dict_ = std::make_unique<Dictionary<Kmer256>>();
+    } else {
+        common::logger->error("k too big: {} > {}", odd_k, 256 / IDictionary::bits_per_char);
+        throw std::runtime_error("k fail");
+    }
+}
+
+DBGSSHash::DBGSSHash(std::string const& input_filename, size_t k, Mode mode)
+      : DBGSSHash(k, mode) {
+    sshash::build_configuration build_config;
+    build_config.k = k;
+    // quick fix for value of m... k/2 but odd
+    build_config.m = std::min(uint64_t((k_ + 1) / 2), sshash::constants::max_m);
```

Comment:
TODO in sshash: `sshash::constants::max_m` should be moved to `kmer_t::max_m`, since we now work with large alphabets. Metagraph should also accommodate to somehow be aware of an actual `kmer_t`?

Reply:
Also if `m` must be odd, it's better to use `(k_ / 2) | 1`, otherwise for `k_ = 7` we'll have `m = 4`.

Reply:
I think it'll be a lot of work to get MetaGraph to work with `kmer_t`, it might be better in the end to have methods in the dictionary class that handle the high-level functionality so we don't have to expose `kmer_t`.

Also, moving `max_m` to `kmer_t` would require making more changes to how sshash computes minimizers. AFAIK, they assume that each m-mer can fit in a uint64_t

Reply:
It still should be moved to `kmer_t`, even under assumption that it fits `uint64_t`, because `max_m` should depend on `bits_per_char`, which is a property of `kmer_t`.